### PR TITLE
enh(csharp) highlight generics in more cases

### PR DESCRIPTION
**Pull request by TupikovVladimir**
https://github.com/highlightjs/highlight.js/issues/2596

### CODE REVIEW

**Comment by yyyc514** [comment]
Tests will also need to be updated:

test/markup/csharp/generic_modifiers*.txt

File: src/languages/csharp.js [hunk 1]
```diff
@@ -177,13 +178,16 @@ export default function(hljs) {
       },
       {
         className: 'function',
-        begin: '(' + TYPE_IDENT_RE + '\\s+)+' + hljs.IDENT_RE + '\\s*\\(', returnBegin: true,
+        begin: '(' + TYPE_IDENT_RE + '\\s+)+' + hljs.IDENT_RE + '\\s*(\\<.+\\>\\s*)?\\(', returnBegin: true,
```
Comment (yyyc514) [inline src/languages/csharp.js:181]: What is the trailing `\\s*` doing?

File: src/languages/csharp.js [hunk 2]
```diff
@@ -177,13 +178,16 @@ export default function(hljs) {
...
         end: /\s*[{;=]/, excludeEnd: true,
         keywords: KEYWORDS,
         contains: [
           {
-            begin: hljs.IDENT_RE + '\\s*\\(', returnBegin: true,
-            contains: [hljs.TITLE_MODE],
+            begin: hljs.IDENT_RE + '\\s*(\\<.+\\>\\s*)?\\(', returnBegin: true,
```
Comment (yyyc514) [inline src/languages/csharp.js:186]: What is the trailing `\\s*` doing?

File: src/languages/csharp.js (hunk 1 above)
Comment (yyyc514) [inline src/languages/csharp.js:181]: I think the trailing `\\s*` should be outside the optional match, yes?

File: src/languages/csharp.js (hunk 2 above)
Comment (yyyc514) [inline src/languages/csharp.js:186]: I think the trailing `\\s*` should be outside the optional match, yes?

File: src/languages/csharp.js (hunk 1 above)
Comment (TupikovVladimir) [reply src/languages/csharp.js:181]: There is no difference in result, but it is better for grouping. Ok, I'll modify it.

**Comment by TupikovVladimir** [comment]
I fixed keywords `in` and `out` for generic parameters, but following white space is included to this keyword. Also I expanded test for generic parameters.

File: src/languages/csharp.js [hunk 3]
```diff
@@ -91,8 +91,11 @@ export default function(hljs) {
   var GENERIC_MODIFIER = {
     begin: "<",
     end: ">",
-    keywords: "in out",
```
Comment (yyyc514) [inline src/languages/csharp.js]: Why was this a problem?

File: src/languages/csharp.js [hunk 4]
```diff
@@ -91,8 +91,11 @@ export default function(hljs) {
...
-    contains: [ TITLE_MODE ]
+    contains: [ 
+      { className: "keyword", begin: "in\\s" },
+      { className: "keyword", begin: "out\\s" },
```
Comment (yyyc514) [inline src/languages/csharp.js:96]: Spaces should NOT be including in keyword tags.

File: src/languages/csharp.js (hunk 3 above)
Comment (TupikovVladimir) [reply src/languages/csharp.js]: Because it does not work, I've made it by another way using keywords, and now space is not  including in keyword tags

Comment (yyyc514) [reply src/languages/csharp.js]: Ah yes, TITLE would eat them...  but the original solution wasn't optimal.  `beginKeywords` is the best quick way to do this!   As you're doing now.

**Comment by yyyc514** [comment]
And while I have your attention should we consider tagging these differently in the future?

title.class vs title.template?  What do you call those items? Do most editors syntax highlight them differently?

We're considering such things for future improvements.

**Pull request merged**